independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/BootstrapAppModelResolver.java

```diff
@@ -68,6 +69,19 @@ public void setBuildTreeLogger(Consumer<String> buildTreeConsumer) {
         this.buildTreeConsumer = buildTreeConsumer;
     }
 
+    /**
+     * Indicates whether application should be resolved to set up the dev mode.
+     * The important difference between the dev mode and the usual build is that
+     * in the dev mode the user application will have to be compiled, so the classpath
+     * will have to include dependencies of scope provided.
```

Comment:
While I can see why it's needed, I'm also wondering if having the provided dependencies in the classpath could lead to a different behavior?

Also wondering if the optional dependencies should be considered? As you will need them to compile too.

Reply:
> While I can see why it's needed, I'm also wondering if having the provided dependencies in the classpath could lead to a different behavior?

It could. Inclusion of provided dependencies may affect the result of the version convergence of the transitive dependencies. While that theoretically could be a big deal, I would say it has the same implications as in case of the usual Maven build (in terms of dependency versions used for the build).

That got me thinking though. `provided` deps in Quarkus apps are going to be much less popular comparing to JEE and other apps that are built to run in some kind of middleware env. Simply because Quarkus build produces a self-contained executable. So nothing is going to provide anything more than what's specified in the pom (including its transitive deps) as required deps.
Quarkus extensions do inject deployment deps with their transitive deps that aren't in the pom but only for the build.
This is very different from building something that is going to be functioning in an environment satisfying your classpath (among other) requirements.
That means the part "you expect the container to provide the dependency at runtime" of the `provided` definition is probably not applicable to Quarkus?
If that's so then `provided` is going to be only a compile time dependency whose presence shouldn't be expected at runtime. What kind of dependencies could that be and could they affect the outcome of Quarkus build? I am thinking that's unlikely. One example (that actually raised this issue) is project Lombock.

To do it properly though, we'd have to re-work the dev mode to create a user app classpath for the compilation and the Quarkus app build classpath in case `provided` deps are found in the pom.

> Also wondering if the optional dependencies should be considered? As you will need them to compile too.

Can you think of a use-case for an optional dep?

Reply:
OK, I think this discussion merits a separate issue if you think there's more progress to do on it.